File: src/AgentBin/Agents/EchoAgent.cs
using A2A;
using A2A.AspNetCore;

namespace AgentBin.Agents;

/// <summary>
/// Simple message-only echo agent. Returns the user's message prefixed with "Echo: ".
/// Validates multi-agent hosting and provides a baseline sanity check for A2A clients.
/// </summary>
public sealed class EchoAgent : IAgentHandler
{
    public async Task ExecuteAsync(RequestContext context, AgentEventQueue eventQueue, CancellationToken cancellationToken)
    {
        var responder = new MessageResponder(eventQueue, context.ContextId);
        var text = context.UserText ?? "(empty message)";
        await responder.ReplyAsync($"Echo: {text}", cancellationToken);
    }

    public static AgentCard GetAgentCard(string agentUrl) =>
        new()
        {
            Name = "Echo Agent",
            Description = "A simple echo agent that returns your message. Use this to verify basic A2A connectivity.",
            Version = "1.0.0",
            SupportedInterfaces =
            [
                new AgentInterface
                {
                    Url = agentUrl,
                    ProtocolBinding = "JSONRPC",
                    ProtocolVersion = "1.0",
                }
            ],
            DefaultInputModes = ["text/plain"],
            DefaultOutputModes = ["text/plain"],
            Capabilities = new AgentCapabilities { Streaming = false, PushNotifications = false },
            Skills =
            [
                new AgentSkill
                {
                    Id = "echo",
                    Name = "Echo",
                    Description = "Echoes back the user message.",
                    Tags = ["echo", "test"],
                }
            ],
        };
}

File: src/AgentBin.Rust/src/echo_agent.rs
use a2a_rs_core::{
    AgentCard, AgentCapabilities, AgentInterface, AgentSkill,
    Message, Part, Role, SendMessageResponse, Task, TaskState, TaskStatus,
    now_iso8601, PROTOCOL_VERSION,
};
use a2a_rs_server::{AuthContext, HandlerResult, MessageHandler};
use async_trait::async_trait;
use uuid::Uuid;

pub struct EchoAgent {
    prefix: String,
}

impl EchoAgent {
    pub fn new(prefix: &str) -> Self {
        Self { prefix: prefix.to_string() }
    }
}

#[async_trait]
impl MessageHandler for EchoAgent {
    async fn handle_message(
        &self,
        message: Message,
        _auth: Option<AuthContext>,
    ) -> HandlerResult<SendMessageResponse> {
        let text = extract_text(&message);
        let context_id = message.context_id.clone().unwrap_or_else(|| Uuid::new_v4().to_string());
        let task_id = Uuid::new_v4().to_string();

        let agent_message = Message {
            kind: "message".to_string(),
            message_id: Uuid::new_v4().to_string(),
            context_id: Some(context_id.clone()),
            task_id: Some(task_id.clone()),
            role: Role::Agent,
            parts: vec![Part::text(format!("Echo: {}", text))],
            metadata: None,
            extensions: vec![],
            reference_task_ids: None,
        };

        let task = Task {
            kind: "task".to_string(),
            id: task_id,
            context_id,
            status: TaskStatus {
                state: TaskState::Completed,
                message: Some(agent_message.clone()),
                timestamp: Some(now_iso8601()),
            },
            history: Some(vec![message, agent_message]),
            artifacts: None,
            metadata: None,
        };

        Ok(SendMessageResponse::Task(task))
    }

    fn agent_card(&self, base_url: &str) -> AgentCard {
        AgentCard {
            name: "AgentBin Echo Agent".to_string(),
            description: "Simple echo agent for basic connectivity testing.".to_string(),
            version: "1.0.0".to_string(),
            supported_interfaces: vec![AgentInterface {
                url: format!("{}{}/v1/rpc", base_url, self.prefix),
                protocol_binding: "JSONRPC".to_string(),
                protocol_version: PROTOCOL_VERSION.to_string(),
                tenant: None,
            }],
            provider: None,
            documentation_url: None,
            capabilities: AgentCapabilities::default(),
            security_schemes: Default::default(),
            security_requirements: vec![],
            default_input_modes: vec!["text".to_string()],
            default_output_modes: vec!["text".to_string()],
            skills: vec![AgentSkill {
                id: "echo".to_string(),
                name: "Echo".to_string(),
                description: "Echoes back the user's message".to_string(),
                tags: vec!["echo".to_string()],
                examples: vec!["hello".to_string()],
                input_modes: vec![],
                output_modes: vec![],
                security_requirements: vec![],
            }],
            signatures: vec![],
            icon_url: None,
        }
    }
}

fn extract_text(message: &Message) -> String {
    message
        .parts
        .iter()
        .filter_map(|p| p.as_text())
        .collect::<Vec<_>>()
        .join("\n")
}
